app/services/funnel/doc_auth/register_campaign.rb

```diff
@@ -2,12 +2,19 @@ module Funnel
   module DocAuth
     class RegisterCampaign
       def self.call(user_id, campaign)
+        return unless campaign_whitelisted?(campaign)
         doc_auth_log = DocAuthLog.find_by(user_id: user_id)
         return if doc_auth_log.nil? || doc_auth_log.no_sp_campaign
         doc_auth_log.no_sp_campaign = campaign.to_s
         doc_auth_log.no_sp_session_started_at = Time.zone.now
         doc_auth_log.save
       end
+
+      def self.campaign_whitelisted?(campaign)
+        whitelist = JSON.parse(Figaro.env.no_sp_campaigns_whitelist || '[]')
+        whitelist.include?(campaign)
+      end
+      private_class_method :campaign_whitelisted?
```

Comment:
personally I think marking methods as private like this is silly since you can just `send/class_eval` to get them anyways...and it prevents unit testing directly if we wanted to

Reply:
Speaking of unit tests, what about adding unit-level tests for this file? They're faster than feature tests and easier to find.

Reply:
Not following wrt `send` since it can be used on instance methods too (ie does it mean we should never use private?) but I agree the method would probably be better suited as it's own service.  I really just thought this entire link/code was going to be once for the tts rollout.  Also, I thought the mode as a group was to gravitate towards feature specs and unit tests would be nice to haves.

Reply:
yeah my belief is that in ruby, `private` is just a suggestion --- no need to change it here

and I didn't get the memo about gravitating towards feature specs, they're much more fragile and slow IMO, but happy to make note of that for the future

Reply:
You and @jmhooper are the spec masters so I'll defer to you guys.  Hoops has done a ton of work overhauling the tests.